### PR TITLE
Feature: ability to add custom auth providers in plugin

### PR DESCRIPTION
**Pull request by EloB**
### What does it do?

I've added possibility to change fields, schema and grant configs. Also added a example plugin with a new Auth provider for Apple in `./examples/getstarted`.

Created separate commits to easily overview what I've done.

These commits can be removed if we don't want that Apple plugin in the codebase. I've added it there for testing purposes. I will put that into a separate plugin on NPM if this PR will be merged.
[759f4600b9097ab547c604e4e6ecb8ae10647f0f](https://github.com/strapi/strapi/pull/15559/commits/759f4600b9097ab547c604e4e6ecb8ae10647f0f)
[0f6a335245e7d0d4b8bb87a700e17a8fbe6397fb](https://github.com/strapi/strapi/pull/15559/commits/0f6a335245e7d0d4b8bb87a700e17a8fbe6397fb)

With these changes it will be able to create isolated auth provider plugins with custom fields etc.
image

### Why is it needed?

Right now it's not possible to support more advanced Auth providers like Apple OAuth without overriding files in users-permissions plugin.

### How to test it?

This is just a draft. If the team likes the PR than we can discuss how to test it.

### Related issue(s)/PR(s)

I've created this PR and it was closed. This PR is taking an alternative route by changes to the core so I can create a plugin that contains Apple Provider and than publish that to the marketplace and maintain that myself.
https://github.com/strapi/strapi/pull/15522

### CODE REVIEW

File: examples/getstarted/src/plugins/apple/server/register.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        },
+      }),
+      buildRedirectUri: ({ urlJoin, baseURL, apiPrefix }) => urlJoin(baseURL, 'apple', 'callback'),
+      connectGrant: (config) => ({
+        ...config,
+        secret: strapi.service('plugin::apple.auth').createSecret(config),
+      }),
+    }
+  );
+};
```
Comment (EloB) [inline examples/getstarted/src/plugins/apple/server/register.js:46]: To create a plugin with a Auth provider you use the `register` lifecycle and register. It now also contains more advanced settings.

File: examples/getstarted/src/plugins/apple/admin/src/index.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        ),
+      };
+    });
+    app.registerPlugin({
+      id: pluginId,
+      initializer: Initializer,
+      isReady: false,
+      name,
+    });
+  },
```
Comment (EloB) [inline examples/getstarted/src/plugins/apple/admin/src/index.js:87]: Inside your plugin admin section you can now also register `transformFields` to make changes to that Modal form. Important that you do those changes immutable.

Comment (EloB) [reply examples/getstarted/src/plugins/apple/admin/src/index.js:87]: This also support changing the schema but this example is not using it.

**Comment by derrickmehaffy** [comment]
I do like this idea as a temporary way to get around the slightly more complex extension process (until we refactor/rebuild U&P and move away from Grant to probably passport instead)

**Comment by Marc-Roig** [comment]
Awesome initiative :) We will discuss this internally and assign someone to review the PR , in the meantime, can you take a look at the tests?

**Comment by EloB** [comment]
@Marc-Roig thanks mate! I can take a look at that. Wasn't sure if you guys were going to like it.

Trying to get familiar with the codebase.

**Comment by Boegie19** [comment]
@EloB If you have any questions about anything or want help with anything best place to get a quick answer is the discord in strapis readme where we also have open office hours every work day with strapi employs every Mon - Fri 12:30 CST

**Comment by EloB** [comment]
@Boegie19 I’ve joined Discord. Will try to join a call on Discord 🙂

**Comment by EloB** [comment]
@derrickmehaffy I also prefer Passport over Grant and totally agree that this is only a temporary solution but yet needed one until someone have time to refactor the Auth beast :D.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy I also prefer Passport over Grant and totally agree that this is only a temporary solution but yet needed one until someone have time to refactor the Auth beast :D.

Yeah we still aren't 100% sure what the plan is but the two proposed options were:

- refactor u&p from the ground up when we work on Strapi v5 at the end of 2023-ish
- build an entirely new plugin and slowly phase out u&p (but both would exist in parallel for a time)

We are probably leaning towards option 2 so in that we would make sure to use passport instead and build a provider system much like our email and upload plugins use.

**Comment by Marc-Roig** [comment]
Hello @EloB , this work is awesome and we have been reviewing both this PR and https://github.com/strapi/strapi/pull/15522. Congrats on doing this.


We have a couple of thoughts, and as Derrick said in the future we might build a new plugin entirely. But this could be a quick win in the meantime, to allow people to use custom providers.

Revising the PR, we think we should take over the work and we want to follow this direction:
- Instead of building a plugin for your provider, it could be done with the provider pattern. The same as in 'upload', where you have the cloudinary, s3-bucket,... providers, we could have the same here.

This would allow a more clear configuration:
```js
// /plugins.js
module.exports = () => {
	permissions: {
		config: {
			provider: 'apple',
			providerOptions: ...
		}
	}
}
```

By looking at what's necessary , the provider should incorporate:
- A way to extend the grantConfig
	-  A method where you can generate secrets as the `createAppleSecret` in your PR
- A way to extend the provider-registry handler
	- And ensure it always return username and email 

Also, as we do not have keys for the apple provider, we could not test this PR :) But we will use a similar structure as yours for implementing this.

We will make a formal proposal for this and start working on it if everyone agrees :)

**Comment by Boegie19** [comment]
Any updates on this?

**Comment by alexandrebodin** [comment]
Hi, We are going to introduce this in v5 only. Here is the work in progress https://github.com/strapi/strapi/pull/20634. Feel free to try it out in the upcoming v5 betas/release candidates and share feedbacks

**Pull request closed without merging**